### DISCUSSION TITLE
Introduce a filter hook mechanism for engine.schema_cache

### DISCUSSION THREAD

**Discussion by Xiaoming-lw** [littleK0i/SnowDDL#131] [Ideas]
As the discussion in [messy environment management ](https://github.com/littleK0i/SnowDDL/discussions/123), we are facing a difficult situation to bring SnowDDL into current production sites.
1. There are a huge number of objects inside current Snowflake (O(1000) for databases, O(1000*100) for schemas), it will take a long time to run SnowDDL, by default 8 workers, would be 20+ minutes for "database,schema".
2. There are dynamic and temporary objects.
3. Most of these databases are shared by different teams, and it is hard to change the current hierarchy of roles and privileges. We don't know what potential issues would be occurred.

In order to skip the unnecessary databases and focus on dozens of databases in the beginning phase, we write customized configs and dump a list of the unnecessary databases. The `__custom/*.py` can make sure we only match the structures of the small number databases.
 
But In current design of SnowDDL, even we restrict the objects types to "database,schema,table", it will still walk through all the tables include which we try to skip in `__custom/*.py`. That is loaded in the `engine.schema_cache`. It would be easy to add some if statements into the for loop to filter result of `show database like ...`. But it is invasive and hard to maintain.

And also there is a discussion in [flexible object filtering](https://github.com/littleK0i/SnowDDL/discussions/90), an "expression" is introduced to include/exclude specific databases. It would help, but still has limitations to more complicated cases, like pattern match for "database/schema/table" name. 

So how about we introduce a filter hook mechanism in the load phase of `schema_cache`, to make it possible to programmatically customize the process of loading objects from Snowflake?

**Comment by littleK0i** [comment]
20 minutes is still quite a lot. SnowDDL should be generally faster than other tools due to built-in parallelism.

Could you run SnowDDL with `--show-timers` CLI option? It should help us to understand which phase takes the most time. This option requires an update to `0.32.0`.

Did you try to increase `--max-workers` to 16 or more? By default it is 8. This option directly affects number of worker threads.

Can you potentially use separate owner role to stop SnowDDL from reading objects from other databases? SnowDDL should skip all databases which are not owned by current role. It also includes everything inside databases: schemas and schema objects. Code: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/cache/schema_cache.py#L30-L31

With `__custom/*.py`, do you call `config.remove_blueprint()` to remove blueprints from config entirely?

---

The main problem with filtering is related to objects being tightly connected with each other. For example, schema roles are required for grants. But in order to create schema roles we need schemas. If some schemas are skipped, roles and grants should be skipped as well. It would be relatively hard to implement filtering feature reliably outside of most basic use cases.

Also, people will start to rely on idea of "applying config partially". Which naturally leads to the situation when parts of config are incomplete or broken, and it is not obvious.

Let's see if we can figure out something else.

---

Btw, I might be available for consulting starting from ~21 Oct. Converting everything automatically and cross-validating with a separate test Snowflake account might be easier than doing slow & painful step-by-step migration for months.

**Reply by Xiaoming-lw** [reply]
Looks like I made a wrong estimation, the 20+ minutes is for objects type "database,schema".
```
2024-10-11 15:17:31.769 - INFO - Executed 0 DDL queries, Suggested 34 DDL queries
2024-10-11 15:17:31.769 - INFO - Timer [InitConfig] elapsed time is 57.068s
2024-10-11 15:17:31.769 - INFO - Timer [InitEngine] elapsed time is 346.142s
2024-10-11 15:17:31.769 - INFO - Timer [AccountParameterResolver] elapsed time is 0.000s
2024-10-11 15:17:31.769 - INFO - Timer [NetworkPolicyResolver] elapsed time is 0.000s
2024-10-11 15:17:31.769 - INFO - Timer [ResourceMonitorResolver] elapsed time is 0.000s
2024-10-11 15:17:31.769 - INFO - Timer [WarehouseResolver] elapsed time is 0.000s
2024-10-11 15:17:31.770 - INFO - Timer [WarehouseRoleResolver] elapsed time is 0.000s
2024-10-11 15:17:31.770 - INFO - Timer [DatabaseResolver] elapsed time is 592.721s
2024-10-11 15:17:31.770 - INFO - Timer [SchemaResolver] elapsed time is 430.532s
2024-10-11 15:17:31.770 - INFO - Timer [ShareRoleResolver] elapsed time is 0.274s
2024-10-11 15:17:31.770 - INFO - Timer [DatabaseRoleResolver] elapsed time is 0.000s
2024-10-11 15:17:31.770 - INFO - Timer [SchemaRoleResolver] elapsed time is 0.000s
2024-10-11 15:17:31.770 - INFO - Timer [SecretResolver] elapsed time is 0.000s
```

For objects type "database,schema,table", it's interrupted by the temporary schema which was removed during the executing of snowddl for more than one hour:
```
2024-10-11 13:46:34.115266, Starting handler
...
2024-10-11 13:53:14.896 - INFO - Snowflake version = 8.38.3 (ENTERPRISE), SnowDDL version = 0.32.0
...
2024-10-11 13:53:15.102 - INFO - Resolved DATABASE [xxxxx]: DROP
...
2024-10-11 13:58:52.107 - INFO - Resolved SCHEMA [xxx]: DROP
2024-10-11 14:51:59.820 - INFO - Could not get existing objects for resolver [TableResolver]:
(
    message   =>  SQL compilation error:
Object does not exist, or operation cannot be performed.
    errno     =>  2043
    sqlstate  =>  02000
    sfqid     =>  01b7a07f-0a0d-2f8a-0001-ad8560cdd08f
    sql       =>  SHOW TABLES IN SCHEMA "xxx"."xxxx"
)
```

Increasing the --max-workers would be help for shorten the running time, but it still failed after 10 minutes running, due to same reason of missing schema.

**Reply by Xiaoming-lw** [reply]
I think the "filter" should be an advanced feature, and it is not an open-box tools for users. So developer should know the dependency inside objects which need to be manipulated, and take them into account.

It seems to be a solution that transfer the ownership of specified dozens of databases to another role and test in a isolated "testing" account. But we still need to face the production account. It's hard to do the ownership transfer against huge data and the running services above them.

**Reply by littleK0i** [reply]
About 6 minutes for `InitEngine` alone is quite long. Internally it runs only `SHOW DATABASES` and `SHOW SCHEMAS IN DATABASE` commands with no additional logic. Should be lightweight.

Do you have anything which may cause very long network round-trip time to Snowflake servers by any chance?

For example:
- running SnowDDL from laptop on 3G connection;
- corporate VPN / PrivateLink;
- network packet monitors (e.g. CarbonBlack);
- being located very far from Snowflake account region;

Even if we optimise this part somehow, role management will be a few magnitudes slower. For every schema 3 roles will be created automatically, with lots of future grants.

Also, SnowDDL assumes schemas will not be deleted during execution. This is very unusual.

**Reply by Xiaoming-lw** [reply]
Hi @littleK0i , 

I think the reason is simply the large number of databases and schemas. Given a total of 500000 objects, each `show` query takes 100ms, and with 100 threads running in parallel, it will take 500s to run snowddl each time.

Considering the difficulty and time required to split and transfer the current roles, it is more direct and easier to skip these unnecessary objects in `schema_cache`. 

Here is a prototype for introducing filter hooks to SnowDDL, would you please take a look at it. If it is a potential possible solution, I would like to complete and contribute it.

https://github.com/littleK0i/SnowDDL/compare/master...Xiaoming-lw:SnowDDL:custom_filter

**Reply by littleK0i** [reply]
@Xiaoming-lw , since you have a very unusual and niche use-case, the best approach might be to:

1. Clone repo and put it into your code base, treating it as internal library.
2. Make changes directly inside the code.

For example, you may add filters logic directly into schema cache code. I expect a lot more changes to come as you keep going. With so many objects and specific requirements it probably makes the most sense.

---

We probably should not add such filters into main lib code, since it would break a lot of things for normal usage, such as:

- Cross-references between schemas, when objects in one schema refers to objects in another schema, but one of schemas are excluded.
- Schema roles and grants for schemas which are excluded are likely going to fail.
- SnowDDL is still going to try and create objects in excluded schemas. Currently the code assumes that all schemas mentioned in config are going to exist and properly configured before resolving any schema objects.

**Reply by Xiaoming-lw** [reply]
Thank you @littleK0i for your suggestion. I understand what you are concerning, and we will keep the experiment in the cloned repo to fit for the use case.
